### PR TITLE
feat : add daily challenge algo, completion,battleTV

### PR DESCRIPTION
**Pull request by shubham-1806**
## In this PR
- Added functionality for /dc/submit route
- Added Score Algorithm for dailychallenge
- Added BattleTV Support for DailyChallenge
- Added tests for DailyChallenge
- Updated APIConfig to support multiple language questions
- Updated aerial attributes of Attackers and Defenders to int
- Updated versions for embedded mongo to support ubuntu 22.04

### CODE REVIEW

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        val coinsLeftPercentage = ((totalCoins - playerCoinsUsed.toDouble()) / totalCoins) * 100
+        return (
+            (
+                getPlayerBaseScore(
+                    coinsLeftPercentage, playerDestruction, perfectBasePart, challType
+                ) +
+                    getTimeScore(perfectTimePart)
+                ) * 100.0
+            )
+            .toInt() / 100.0
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt:68]: Is this toInt() needed ?

File: server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
 
 @Service
 class DailyChallengeService(
-    @Autowired private val dailyChallengeRepository: DailyChallengeRepository
+    @Autowired private val dailyChallengeRepository: DailyChallengeRepository,
+    @Autowired private val publicUserService: PublicUserService,
+    @Autowired private val dailyChallengeScoreAlgorithm: DailyChallengeScoreAlgorithm
 ) {
 
     @Value("\${environment.event-start-date}") val tempDate: String = ""
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt]:
```suggestion
    @Value("\${environment.event-start-date}") private lateinit var currentDate: String
```

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import delta.codecharacter.server.config.GameConfiguration
+import delta.codecharacter.server.daily_challenge.DailyChallengeEntity
+import org.springframework.beans.factory.annotation.Value
+import java.time.Duration
+import java.time.Instant
+import kotlin.math.exp
+
+class DailyChallengeScoreAlgorithm : ScoreAlgorithm {
+
+    @Value("\${environment.event-start-date}") val tempDate: String = ""
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt]:
```suggestion
    @Value("\${environment.event-start-date}") private lateinit var tempDate: String
```

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        return perfectTimeScore * exp((-1) * (hours / 15))
+    }
+
+    override fun getDailyChallengeScore(
+        playerCoinsUsed: Int,
+        playerDestruction: Double,
+        dailyChallenge: DailyChallengeEntity
+    ): Double {
+        val gameConfiguration = GameConfiguration()
+        val totalCoins = gameConfiguration.gameParameters().numberOfCoins
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt]: You don't have to create an `instance` and call again. `GameParameter` is already a `bean` just autowire and inject it

File: server/src/main/kotlin/delta/codecharacter/server/user/public_user/PublicUserEntity.kt [hunk 5]
```diff
@@ -22,5 +22,5 @@ data class PublicUserEntity(
     val ties: Int,
     val isActivated: Boolean = true,
     val score: Double,
-    val isDailyChallengeCompleted: Boolean = false,
+    val dailyChallengeHistory: HashMap<UUID, DailyChallengeHistory>
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/user/public_user/PublicUserEntity.kt]: It should be a `HashMap` of  Day number and a `Dataclass` of *Score*,*DailychallengeEntity*

File: server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt [hunk 6]
```diff
@@ -23,34 +31,45 @@ class DailyChallengeService(
...
-     *
-     * Corresponding LeaderBoard updates
-     *
-     * Scheduling for isDailyChallengeComplete -> false by 24hrs
-     */
+    fun completeDailyChallenge(gameEntity: GameEntity, userId: UUID): DailyChallengeMatchVerdictEnum {
+        val (_, coinsUsed, destruction, _, _) = gameEntity
+        if (gameEntity.status == GameStatusEnum.EXECUTE_ERROR)
+            return DailyChallengeMatchVerdictEnum.FAILURE
+        val dc = dailyChallengeRepository.findByDay(findNumberOfDays()).get()
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt]: Instead you can make use of the `getDailyChallengeByDate()` function already declared in `DailyChallengeService.kt` .

File: server/src/main/kotlin/delta/codecharacter/server/user/public_user/DailyChallengeHistory.kt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+package delta.codecharacter.server.user.public_user
+
+data class DailyChallengeHistory(val score: Double, val day: Int)
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/user/public_user/DailyChallengeHistory.kt]:
```suggestion
data class DailyChallengeHistory(val score: Double, val challenge: DailyChallengeEntity)
```

File: server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt [hunk 8]
```diff
@@ -312,17 +316,37 @@ class MatchService(
...
+                    match.copy(
+                        verdict =
+                        dailyChallengeService.completeDailyChallenge(updatedGame, match.user.userId)
+                    )
+                notificationService.sendNotification(
+                    match.user.userId,
+                    title = "Daily Challenge Results",
+                    content =
+                    when (updatedMatch.verdict) {
+                        DailyChallengeMatchVerdictEnum.SUCCESS -> "Successfully Completed Challenge"
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt]:
```suggestion
                        DailyChallengeMatchVerdictEnum.SUCCESS -> "Successfully completed challenge"
```

File: server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt [hunk 9]
```diff
@@ -312,17 +316,37 @@ class MatchService(
...
+                        else -> {
+                            "Failed to Complete Challenge"
+                        }
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt]:
```suggestion
                        else -> {
                            "Failed to complete challenge"
                        }
```

File: server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt [hunk 10]
```diff
@@ -142,10 +142,14 @@ class MatchService(
...
     fun createDCMatch(userId: UUID, dailyChallengeMatchRequestDto: DailyChallengeMatchRequestDto) {
-        val (_, chall, challType, _, completionStatus) = dailyChallengeService.getDailyChallengeByDate()
+        val (_, chall, challType, _, completionStatus) =
+            dailyChallengeService.getDailyChallengeByDateForUser(userId)
+        val dc = dailyChallengeService.getDailyChallengeByDate()
         val (value, _) = dailyChallengeMatchRequestDto
         if (completionStatus != null && completionStatus) {
-            throw CustomException(HttpStatus.ACCEPTED, "You have already completed your daily Challenge")
+            throw CustomException(
+                HttpStatus.BAD_REQUEST, "You have already completed your daily Challenge"
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt]:
```suggestion
                HttpStatus.BAD_REQUEST, "You have already completed your daily challenge"
```

**Review by Ram-20062003 — CHANGES_REQUESTED** [review]
- [x] Have `map validator` check for `CODE AS QUESTION` challenges
- [x] Have better meaningful names for the variable 
- [x] Revert back the deleted `SchedulingService` file.
- [x] Add `BattleTV` support also for `DCs`.
- [x] Look into grammatical errors also:sneezing_face:

File: server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt [hunk 11]
```diff
@@ -188,6 +196,9 @@ class MatchService(
                 }
                 createDualMatch(userId, createMatchRequestDto.opponentUsername!!)
             }
+            else -> {
+                throw IllegalStateException("MatchMode Is Not Correct")
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt]:
```suggestion
                throw CustomExeption(HttpStatus.BAD_REQUEST,"MatchMode Is Not Correct")
```

File: server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeController.kt [hunk 12]
```diff
@@ -21,7 +21,9 @@ class DailyChallengeController(
 ) : DailyChallengesApi {
     @Secured(value = ["ROLE_USER"])
     override fun getDailyChallenge(): ResponseEntity<DailyChallengeGetRequestDto> {
-        return ResponseEntity.ok(dailyChallengeService.getDailyChallengeByDate())
+        val user = SecurityContextHolder.getContext().authentication.principal as UserEntity
+        val userId = user.id
+        return ResponseEntity.ok(dailyChallengeService.getDailyChallengeByDateForUser(userId))
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeController.kt]:
```suggestion
        return ResponseEntity.ok(dailyChallengeService.getDailyChallengeByDateForUser(user.id))
```

File: server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt [hunk 13]
```diff
@@ -1,56 +1,75 @@
...
         val period: Duration = Duration.between(givenDateTime, nowDateTime)
         return period.toDays().toInt()
     }
 
-    fun getDailyChallengeByDate(): DailyChallengeGetRequestDto {
-
+    fun getDailyChallengeByDate(): DailyChallengeEntity {
         val dc =
             dailyChallengeRepository.findByDay(findNumberOfDays()).orElseThrow {
                 throw CustomException(org.springframework.http.HttpStatus.BAD_REQUEST, "Invalid Request")
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt]:
```suggestion
        return dailyChallengeRepository.findByDay(findNumberOfDays()).orElseThrow {
                throw CustomException(HttpStatus.BAD_REQUEST, "Invalid day for daily challenge")
            }
```

File: server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt [hunk 14]
```diff
@@ -1,56 +1,75 @@
...
-     *
-     * Corresponding LeaderBoard updates
-     *
-     * Scheduling for isDailyChallengeComplete -> false by 24hrs
-     */
+    fun completeDailyChallenge(gameEntity: GameEntity, userId: UUID): DailyChallengeMatchVerdictEnum {
+        val (_, coinsUsed, destruction, _, _) = gameEntity
+        if (gameEntity.status == GameStatusEnum.EXECUTE_ERROR)
+            return DailyChallengeMatchVerdictEnum.FAILURE
+        val dc = getDailyChallengeByDate()
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt]: `dc` -> `currentDailyChallenge` everywhere

File: server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt [hunk 15]
```diff
@@ -1,56 +1,75 @@
...
+        if ((dc.challType == ChallengeTypeDto.MAP && destruction > dc.toleratedDestruction) ||
+            (dc.challType == ChallengeTypeDto.CODE && destruction < dc.toleratedDestruction)
+        ) {
+            val score =
+                dailyChallengeScoreAlgorithm.getDailyChallengeScore(
+                    playerCoinsUsed = coinsUsed,
+                    playerDestruction = destruction,
+                    dailyChallenge = dc,
+                )
+            val updatedDc = dc.copy(numberOfCompletions = dc.numberOfCompletions + 1)
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/daily_challenge/DailyChallengeService.kt]: `updatedDc` -> `updatedDailyChallenge` everywhere

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt]: Please use `gameParameter` bean here

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        val (_, _, _, challType, _, _, perfectScore, numberOfCompletions) = dailyChallenge
+        val perfectBasePart = 0.7 * perfectScore * exp(((-1) * (numberOfCompletions.toDouble() / 150)))
+        val perfectTimePart = 0.3 * perfectScore
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt]: Declare all constants (0.7,0.3,150 ) in separate variables and use variables here.Give proper naming for all constants too.
For example,
- private val perfectBaseScoreConstant = 0.7
- private val perfectTimeScoreConstant = 0.3

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        destructionPercent: Double,
+        perfectBaseScore: Double,
+        challType: ChallengeTypeDto
+    ): Double {
+        if (challType == ChallengeTypeDto.CODE)
+            return ((100.0 - coinsLeftPercent) + (2 * (100 - destructionPercent)) + perfectBaseScore)
+        return (coinsLeftPercent + (2 * destructionPercent) + perfectBaseScore)
+    }
+
+    override fun getTimeScore(perfectTimeScore: Double): Double {
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt]: `getTimeScore` -> `getPlayerTimeScore`

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+        val hours = getHoursSinceDailyChallengeLaunched()
+        return perfectTimeScore * exp((-1) * (hours / 15))
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/DailyChallengeScoreAlgorithm.kt]: Declare constant (15) in variable and use that variable here as said earlier.

File: server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/ScoreAlgorithm.kt [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+    fun getPlayerBaseScore(
+        coinsLeftPercent: Double,
+        destructionPercent: Double,
+        perfectBaseScore: Double,
+        challType: ChallengeTypeDto
+    ): Double
+
+    fun getHoursSinceDailyChallengeLaunched(): Double
+
+    fun getTimeScore(
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/logic/daily_challenge_score/ScoreAlgorithm.kt]: `getTimeScore` -> `getPlayerTimeScore`

File: server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt [hunk 21]
```diff
@@ -312,17 +359,37 @@ class MatchService(
...
+                notificationService.sendNotification(
+                    match.user.userId,
+                    title = "Daily Challenge Results",
+                    content =
+                    when (updatedMatch.verdict) {
+                        DailyChallengeMatchVerdictEnum.SUCCESS -> "Successfully completed challenge"
+                        else -> {
+                            "Failed to complete challenge"
+                        }
+                    }
```
Comment (Ram-20062003) [inline server/src/main/kotlin/delta/codecharacter/server/match/MatchService.kt]: Add `DailyChallengeMatchVerdictEnum.FAILURE` separately and in `else` mention `Some error occured. Try again!`.

**Review by Ram-20062003 — CHANGES_REQUESTED** [review]

**Comment by Ram-20062003** [comment]
Please, add the changes done in this PR in description.

**Review by Ram-20062003 — APPROVED** [review]

**Review by vigneshd332 — APPROVED** [review]

**Pull request merged**